[NP-5449] Session OrX

jlhughes opened the pull request:
Each client call, on each request, whether DIG, HTTP, Socket,
performs session context setup via applyTo. applyTo is the most
expensive logic of context setup, but the results of applyTo do
not change between calls, so it makes sense to reuse the
context setup. Auth checks are independent of this.
Also, if the system were to reconfigure a theme, for example,
the existing sessions could be invalidated or the the 'context'
of each be reset to force applyTo to run on next client call.

This change in a medusa cluster provides a 20% performance gain
for 'ping' test on a 1Gbs network.

jlhughes committed:
Implement Or X and use it to merge request and cached session context

jlhughes committed:
Implemented caching of the configuration of the context during Session.applyTo.

On subsequent calls, the expensive theme, appConfig, ... setup is
reused.
The use of the cached context is via an Or X which combines the argument
context and the cached context; looking up in both but only putting to
the 'delegate'.

jlhughes committed:
remove logging of session id

jlhughes committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into NP-5449/SessionOrX

kgrgreer commented on src/foam/core/OrX.java:
```diff
@@ -0,0 +1,54 @@
+  }
+
+  public <T> T get(Class<T> key) {
+    T t = getX().get(key);
```
I think this should be getX().get(this, key);

kgrgreer commented on src/foam/core/OrX.java:
```diff
@@ -0,0 +1,54 @@
+
+  public <T> T get(Class<T> key) {
+    T t = getX().get(key);
+    if ( t == null ) return localX_.get(key);
```
localX_.get(this, key);

kgrgreer commented on src/foam/core/OrX.java:
```diff
@@ -0,0 +1,54 @@
+
+  public int getInt(X x, Object key, int defaultValue) {
+    Object o = getX().getInt(x, key, defaultValue);
+    if ( o == null ) return (int) localX_.get(key);
```
Will this ever happen? Won't it just return the defaultValue? Same for the other typed gets below.

jlhughes committed:
correct logic when dealing with defaultValues

kgrgreer commented on src/foam/core/OrX.java:
```diff
@@ -0,0 +1,56 @@
+    get - first test delegate, then local
+    put - put to delegate
+ **/
+public class OrX
```
I think the put() semantics is incorrect. If you put(), it will go to the delegate, which means you can't override any values from localX_. I think on put() you would need to do something like:
localX_ = localX_.put().

jlhughes replied:
The OR'd contexts are like a linked list. 
localX -> delegate -> delegate
get - lookup in delegate first, working it's way back to localX
put - puts into delegate
In all my use cases localX is the requestContext or some other root context.

jlhughes replied:
ah, my initial version and documentation was reversed.

jlhughes committed:
Merge branch 'development' of github.com:kgrgreer/foam3 into NP-5449/SessionOrX

jlhughes committed:
implement cd

jlhughes committed:
Dump stacktrace of any NPEs

jlhughes committed:
Add additional logic around testing subject to avoid NPEs.

Output first part of Session id in log messages. Session id is a
long running id which can be used for tracing.a

jlhughes committed:
correct test for cached user
